File: client-scalaz/src/main/scala/org/openapitools/client/api/Album.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import Album._

case class Album (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
latitude: Option[Double],
longitude: Option[Double],
locationDescription: Option[String],
owner: Option[Account],
title: Option[String],
application: Option[Application],
approvalStatus: Option[ApprovalStatus],
publicPermissions: Option[Permissions],
userPermissions: Option[List[UserPermissions]],
visibility: Option[Visibility],
notes: Option[List[Note]],
noteCount: Option[Long],
likes: Option[List[YayOrNay]],
likeCount: Option[Long],
dislikeCount: Option[Long],
flags: Option[List[Flag]],
flagCount: Option[Long],
flagThreshold: Option[FlagThreshold],
description: Option[String],
albumType: Option[String],
assets: Option[List[Asset]],
coverAsset: Option[Asset],
tags: Option[String],
startDate: Option[OffsetDateTime],
endDate: Option[OffsetDateTime],
contactInfo: Option[ContactInfo],
assetCount: Option[Long],
favoriteCount: Option[Long],
viewedCount: Option[Long],
clickedCount: Option[Long],
sharedCount: Option[Long],
anonymous: Option[Boolean],
ownerUsername: Option[String],
display: Option[String],
appKey: Option[String],
billableId: Option[Long],
subType: Option[String],
contentAsset: Option[Asset],
secondaryType: Option[String],
contentName: Option[String],
applicationTitle: Option[String],
shortTitle: Option[String],
defaultThreshold: Option[Long])

object Album {
  import DateTimeCodecs._
  sealed trait ApprovalStatus
  case object PENDING extends ApprovalStatus
  case object REJECTED extends ApprovalStatus
  case object APPROVED extends ApprovalStatus
  case object FEATURED extends ApprovalStatus

  object ApprovalStatus {
    def toApprovalStatus(s: String): Option[ApprovalStatus] = s match {
      case "PENDING" => Some(PENDING)
      case "REJECTED" => Some(REJECTED)
      case "APPROVED" => Some(APPROVED)
      case "FEATURED" => Some(FEATURED)
      case _ => None
    }

    def fromApprovalStatus(x: ApprovalStatus): String = x match {
      case PENDING => "PENDING"
      case REJECTED => "REJECTED"
      case APPROVED => "APPROVED"
      case FEATURED => "FEATURED"
    }
  }

  implicit val ApprovalStatusEnumEncoder: EncodeJson[ApprovalStatus] =
    EncodeJson[ApprovalStatus](is => StringEncodeJson(ApprovalStatus.fromApprovalStatus(is)))

  implicit val ApprovalStatusEnumDecoder: DecodeJson[ApprovalStatus] =
    DecodeJson.optionDecoder[ApprovalStatus](n => n.string.flatMap(jStr => ApprovalStatus.toApprovalStatus(jStr)), "ApprovalStatus failed to de-serialize")
  sealed trait Visibility
  case object PUBLIC extends Visibility
  case object `PRIVATE` extends Visibility
  case object FRIENDS extends Visibility

  object Visibility {
    def toVisibility(s: String): Option[Visibility] = s match {
      case "PUBLIC" => Some(PUBLIC)
      case "`PRIVATE`" => Some(`PRIVATE`)
      case "FRIENDS" => Some(FRIENDS)
      case _ => None
    }

    def fromVisibility(x: Visibility): String = x match {
      case PUBLIC => "PUBLIC"
      case `PRIVATE` => "`PRIVATE`"
      case FRIENDS => "FRIENDS"
    }
  }

  implicit val VisibilityEnumEncoder: EncodeJson[Visibility] =
    EncodeJson[Visibility](is => StringEncodeJson(Visibility.fromVisibility(is)))

  implicit val VisibilityEnumDecoder: DecodeJson[Visibility] =
    DecodeJson.optionDecoder[Visibility](n => n.string.flatMap(jStr => Visibility.toVisibility(jStr)), "Visibility failed to de-serialize")

  implicit val AlbumCodecJson: CodecJson[Album] = CodecJson.derive[Album]
  implicit val AlbumDecoder: EntityDecoder[Album] = jsonOf[Album]
  implicit val AlbumEncoder: EntityEncoder[Album] = jsonEncoderOf[Album]
}

File: client-clojure/src/sirqul_io_t_platform/specs/album.clj
(ns sirqul-io-t-platform.specs.album
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.application :refer :all]
            [sirqul-io-t-platform.specs.permissions :refer :all]
            [sirqul-io-t-platform.specs.user-permissions :refer :all]
            [sirqul-io-t-platform.specs.note :refer :all]
            [sirqul-io-t-platform.specs.yay-or-nay :refer :all]
            [sirqul-io-t-platform.specs.flag :refer :all]
            [sirqul-io-t-platform.specs.flag-threshold :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.contact-info :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            )
  (:import (java.io File)))


(def album-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :latitude) float?
   (ds/opt :longitude) float?
   (ds/opt :locationDescription) string?
   (ds/opt :owner) account-spec
   (ds/opt :title) string?
   (ds/opt :application) application-spec
   (ds/opt :approvalStatus) string?
   (ds/opt :publicPermissions) permissions-spec
   (ds/opt :userPermissions) (s/coll-of user-permissions-spec)
   (ds/opt :visibility) string?
   (ds/opt :notes) (s/coll-of note-spec)
   (ds/opt :noteCount) int?
   (ds/opt :likes) (s/coll-of yay-or-nay-spec)
   (ds/opt :likeCount) int?
   (ds/opt :dislikeCount) int?
   (ds/opt :flags) (s/coll-of flag-spec)
   (ds/opt :flagCount) int?
   (ds/opt :flagThreshold) flag-threshold-spec
   (ds/opt :description) string?
   (ds/opt :albumType) string?
   (ds/opt :assets) (s/coll-of asset-spec)
   (ds/opt :coverAsset) asset-spec
   (ds/opt :tags) string?
   (ds/opt :startDate) inst?
   (ds/opt :endDate) inst?
   (ds/opt :contactInfo) contact-info-spec
   (ds/opt :assetCount) int?
   (ds/opt :favoriteCount) int?
   (ds/opt :viewedCount) int?
   (ds/opt :clickedCount) int?
   (ds/opt :sharedCount) int?
   (ds/opt :anonymous) boolean?
   (ds/opt :ownerUsername) string?
   (ds/opt :display) string?
   (ds/opt :appKey) string?
   (ds/opt :billableId) int?
   (ds/opt :subType) string?
   (ds/opt :contentAsset) asset-spec
   (ds/opt :secondaryType) string?
   (ds/opt :contentName) string?
   (ds/opt :applicationTitle) string?
   (ds/opt :shortTitle) string?
   (ds/opt :defaultThreshold) int?
   })

(def album-spec
  (ds/spec
    {:name ::album
     :spec album-data}))
